### PR TITLE
added content upload modal

### LINKED ISSUE #244: [Content Seeding] Add a page for youtube upload
## Description
Add a page to submit youtube urls to node server to upload videos. This page should be only for internal authorised members. The list can be fetched from the node server.

### PR DESCRIPTION
**Pull request by ravibazz**
## Changes
- added changes for content upload modal.

## Impact 
- fixes #244

### CODE REVIEW

File: src/auth/server_impl/store/redis_kv.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ impl KVStore for RedisKV {
 
     async fn write(&self, key: String, value: String) -> Result<(), KVError> {
         let mut con = self.0.get().await?;
-        con.hset(key, AUTH_FIELD, value).await?;
+        con.hset::<_, _, _, ()>(key, AUTH_FIELD, value).await?;
```
Comment (ravibazz) [inline src/auth/server_impl/store/redis_kv.rs:27]: This to fix the clippy error. Related rust lang issue: https://github.com/rust-lang/rust/issues/123748

File: src/component/content_upload.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
+use crate::{auth::DelegatedIdentityWire, state::canisters::Canisters};
+use leptos::*;
+#[component]
```
Comment (rupansh-gob) [inline src/component/content_upload.rs:3]: nit: probably want to add some spacing here?

File: src/state/canisters.rs [hunk 3]
```diff
@@ -92,6 +98,28 @@ impl Canisters<true> {
...
+
+    pub async fn upload_using_content_seed(
+        &self,
+        delegated_identity: DelegatedIdentityWire,
+        url: String,
+    ) -> Result<(), Box<dyn Error>> {
+        self.content_seed_client
+            .upload_content(url, delegated_identity)
+            .await
+    }
```
Comment (rupansh-gob) [inline src/state/canisters.rs:122]: these functions are unrelated to `Canisters`, content_seed_client should not be in the `Canisters` utility

File: src/state/content_seed_client.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        let res = self.client.get(api_url).send().await?;
+        let res_json: AllowPrincpalRes = res.json().await?;
+        Ok(res_json.allowed)
+    }
+
+    pub async fn upload_content(
+        &self,
+        url: String,
+        identity: DelegatedIdentityWire,
+    ) -> Result<(), Box<dyn Error>> {
```
Comment (rupansh-gob) [inline src/state/content_seed_client.rs:56]: prefer `ServerFnError` instead of `Box<dyn Error>`
More ideally, i'd like fully typed errors

File: src/page/menu.rs [hunk 5]
```diff
@@ -136,49 +144,126 @@ fn NsfwToggle() -> impl IntoView {
...
+                                                            Upload Content
+                                                        </button>
+                                                    </div>
+                                                },
+                                            );
+                                        }
+                                        None
+                                    }}
+
+                                </div>
```
Comment (rupansh-gob) [inline src/page/menu.rs:226]: Create a new component, use `<Show>`
Something like
```
<Show when=move || is_authorized.get().flatten().unwrap_or_default()>
....
</Show>
```

File: src/page/menu.rs [hunk 6]
```diff
@@ -136,49 +144,126 @@ fn NsfwToggle() -> impl IntoView {
...
+
+    let can_res = authenticated_canisters();
+    let is_authorized = create_local_resource_with_initial_value(
+        move || MockPartialEq(can_res()),
+        move |can_res| async move {
+            let canisters = can_res.0?.ok()?;
+            canisters.is_user_authorized_to_seed_content().await.ok()
+        },
+        Some(Some(false)),
+    );
```
Comment (rupansh-gob) [inline src/page/menu.rs:158]: Probably want to cache this value in context (like how we cache posts)

File: src/page/menu.rs [hunk 7]
```diff
@@ -136,49 +144,126 @@ fn NsfwToggle() -> impl IntoView {
...
-                    <div class="flex flex-row justify-center gap-4 items-center px-4">
-                        <ProfileInfo/>
-                    </div>
-                    <Show when=move || !is_connected()>
-                        <div class="w-full px-8 md:w-4/12 xl:w-2/12">
-                            <ConnectLogin/>
-                        </div>
-                        <div class="w-full px-8 text-center text-sm font-sans">
-                            {r#"Your Yral account has been setup. Login with Google to not lose progress."#}
+        <Suspense fallback=move || {
```
Comment (rupansh-gob) [inline src/page/menu.rs:172]: Use a fine grained suspense.
only YoutubeUpload modal should show the loading spinner
UploadContent button can stay hidden till the resource is completed

**Review by rupansh-gob — CHANGES_REQUESTED** [review]

File: src/state/content_seed_client.rs (hunk 4 above)
Comment (ravibazz) [reply src/state/content_seed_client.rs:56]: What additional benefit do we have mentioning ServerFnError we cannot recover from it and we will just print out the message from the error so doesn't make sense to use strongly typed error.

Comment (rupansh-gob) [reply src/state/content_seed_client.rs:56]: The added benefit is consistency, the leptos ecosystem uses ServerFnError for arbitrary errors and so does the code base.
Why does function XYZ use ServerFnError but ABC use Box<dyn Error>?

Strong typing is always preferred, in the future we might have better error monitoring and proper types give free information about what the error is.

Comment (ravibazz) [reply src/state/content_seed_client.rs:56]: Consistency is not an issue here we can move every api to return Box<dyn Error>. Strong typing is preferred if you can actually recover form the error or have some additional logic for error type. Monitoring would log the error as far as I know and using typed erased error when you just want to know the cause is preferred.

Comment (rupansh-gob) [reply src/state/content_seed_client.rs:56]: Consistency is an issue for the code base, I dont see why we'd want ti deviate from what we already do.
Its similar to how a crate that uses anyhow wouldnt use Box<dyn Error>.

Secondly,
You don't know the future and neither do I, We dont know what the monitoring API will look liike, we dont know how we will be handling errors in the future. Strong typing is a defensive mechanism that will cause us less headaches in the future. If you have really strong opinions against strongly typed errors, i dont really mind, we can add the types whenever required, but it has been discussed well before and multiple times that we should prefer strong types in the leptos code base (cc: @s-go-bazzinga @komal-rs )

Comment (ravibazz) [reply src/state/content_seed_client.rs:56]: I don't mind. I just wrote what was suggested by Jon Gjengset in one his book.

_In general, the community consensus is that errors should be rare and therefore should not add much cost to the "happy path." For that reason, errors are often placed behind a pointer type, such as a Box or Arc. This way, they're unlikely to add much to the size of the overall Result type they're contained within._

File: src/page/menu.rs (hunk 7 above)
Comment (ravibazz) [reply src/page/menu.rs]: Done

File: src/page/menu.rs (hunk 6 above)
Comment (ravibazz) [reply src/page/menu.rs]: Done

File: src/state/canisters.rs (hunk 3 above)
Comment (ravibazz) [reply src/state/canisters.rs]: Done

**Review by rupansh-gob — APPROVED** [review]

File: src/state/content_seed_client.rs (hunk 4 above)
Comment (s-go-bazzinga) [reply src/state/content_seed_client.rs:66]: Overall, always prefer strong typing wherever possible. You get meagre storage benefits in terms of lesser sized compile outputs but drastically reduced safety/compile time guarantees. And more CPU usage as well due to dynamic inferencing.

Storage is cheap, compute almost always costlier.

Finally make the compiler do as much as possible. The lesser a dev needs to remember to "jump", the lesser chances they're going to trip..

My 2 cents.

Comment (komal-rs) [reply src/state/content_seed_client.rs:66]: Agreed. 
Based on some searches, looks like this is an active debate in the Rust community with even some performant crates like reqwest and ripgrep going the other route of dyn - https://users.rust-lang.org/t/result-type-when-to-box-vs-when-to-wrap/59084 . 

I think the important thing is for us to agree on one approach and be consistent with it as much as possible, atleast localized to relevant repos.

Comment (ravibazz) [reply src/state/content_seed_client.rs:66]: I think type erased doesn't mean there would be no compile time checks. The only idea is errors are rare cases and if you don't want to handling based on error types and just want to know the cause it is okay to just send `dyn Error` which provides the reason but doesn't provide data that is not needed.

Comment (s-go-bazzinga) [reply src/state/content_seed_client.rs:66]: Chiming in again. Those crates are library crates that will have hundreds of eyeballs validating that there are no errors and millions of dependent crates that are battle testing in production that everything runs as expected. In our case, we don't have that privilege and we wan't to get as many compiler guarantees regarding safety as possible.

Also Boxed generics are ugly as heck for the reader

Comment (ravibazz) [reply src/state/content_seed_client.rs:66]: I am just confused because I don't find a geniuene case over here include types. Box generics in erros help in developer experience by returning any type of error that implements error trait that is the only useful information that is needed as even in error case we are printing the error cause. What benefit does Typed Error type enum/struct provide in these use case over Boxed type. I have two types of error in the same block one retrieved from json and one from request itself. If I use typed I need to return an enum that implements Error but given that I will not be using that anywhere in the code what is the use? Plus `Box<dyn Error>` also gurantees you will only be able to access methods in trait Error which is all that will be needed for logging.

**Review by rupansh-gob — APPROVED** [review]
